policy: Do not scan all entries when adding new ones

jrajahalme opened the pull request:
Hold MapStateEntries in `entries` map, `trie` and `cidr` within the `trie` are additional indices for the same MapStateEntries. `cidr` only indexes keys with CIDR identities, while `trie` indexes keys by traffic direction, protocol, destination port and mask.

Use the `trie` and the new `cidr` index when adding entries to MapState with deny entries, and the existing `trie` index when adding entries to MapState with auth entries. Likewise, use the `trie` index when adding visibility entries. With this we no longer scan through all entries in MapState when adding new entries.

MapState now contains an optional validator interface that can be set in unit tests to add calls to validation code for the new ancestor/descendants code. This is currently enabled only for the `denyPreferredInsertWithChanges`.

On local benchmarking for generating MapState for an egress policy with 1000 deny and 1000 allow CIDR rules:

Before:
```
Number of MapState entries: 117515
goos: linux
goarch: arm64
pkg: github.com/cilium/cilium/pkg/policy
BenchmarkRegenerateCIDRDenyPolicyRules-12    	       1	763361909494 ns/op
PASS
ok  	github.com/cilium/cilium/pkg/policy	763.696s
```
After:
```
Number of MapState entries: 117515
goos: linux
goarch: arm64
pkg: github.com/cilium/cilium/pkg/policy
BenchmarkRegenerateCIDRDenyPolicyRules-12    	Number of MapState entries: 117515
Number of MapState entries: 117515
       3	 418981451 ns/op
PASS
ok  	github.com/cilium/cilium/pkg/policy	3.434s
```
This is a 1820x speedup :-)

Using the PortProto `trie` in addition to the new `cidr` trie as suggested by @nathanjsweet made this 8x faster for this specific test case.

```release-note
Deny rule processing has been improved to run faster, especially on larger policies.
```

jrajahalme committed:
policy: Clarify comment

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: avoid indirection for contained mapStateMaps

Avoid unnecessary indirection.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Avoid copies for allKey

Do not copy 'allKey' when not really needed. Have one of them for each
traffic direction instead.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: add a main mapstate map

Add back the simple Go map to hold MapStateEntries, make trie just an
index to it.

This change allows adding additional indices in later commits.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Generalize bootstrapRepo for different sets of identities

Change bootstrapRepo to not generate identities directly, but via the
passed in rule generation function. This way the generated identities can
be tailored for the rules being used. This becomes useful with additional
CIDR tests.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Remove dead test code

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Refactor newMapState

Remove the initMap from newMapState() and add withState(), and the same
for the interface returning versions (NewMapState() and WithState()). This
helps reducing refactor changes from the next commit.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed commented on pkg/policy/selectorcache.go:
```diff
@@ -474,18 +472,11 @@ func (sc *SelectorCache) UpdateIdentities(added, deleted identity.IdentityMap, w
-// it returns both IPv4 and IPv6.
-func (sc *SelectorCache) GetNetsLocked(id identity.NumericIdentity) []*net.IPNet {
+// GetPrefixLocked returns the most specific CIDR for an identity, if any.
+func (sc *SelectorCache) GetPrefixLocked(id identity.NumericIdentity) netip.Prefix {
```
What if we make a separate `sync.Map[id]prefix`. Then we can drop this lock, which would be a big help while rolling out incremental updates.

jrajahalme replied:
Good point, this would remove the need for the selector cache locking added here.

jrajahalme replied:
Done :-)

jrajahalme replied:
Do you think we could also drop the locking here:
```
// ConsumeMapChanges transfers the changes from MapChanges to the caller,
// locking the selector cache to make sure concurrent identity updates
// have completed.
// PolicyOwner (aka Endpoint) is also locked during this call.
func (p *EndpointPolicy) ConsumeMapChanges() (adds, deletes Keys) {
	p.selectorPolicy.SelectorCache.mutex.RLock()
	defer p.selectorPolicy.SelectorCache.mutex.RUnlock()
```
From the comment it seems possible that the mutex is needed also to make sure an identity that is being added is actually done before we proceed here. So while `sync.Map()` is safe, if we drop this lock it could happen that the prefix is not yet found when we ask for it?

Or do we have some other synchronization that takes care of the ordering of events here?

jrajahalme replied:
Note that after the last commit the access pattern to `SelectorCache.GetPrefix()` is different in that it is called for the added entry only, as everything else is already in the index.

For incremental changes it is the new identity that was just added that is of interest, though.

jrajahalme replied:
It seems the Prefix is now added to the `prefixMap` strictly before incremental updates are created, but maybe we need to be synchronizing for the concurrent updates to have propagated to the MapChanges so that we get them? Without this lock it could be possible that we "used" the trigger too early and just barely missed some of the updates, and we're not going to be triggered again until some periodic map sync etc.?

jrajahalme replied:
We are explicitly handing this part through a waitgroup before we call endpoint's ApplyPolicyMapChanges:
```
			notifyWg.Wait()
			if err := ep.ApplyPolicyMapChanges(proxyWaitGroup); err != nil && !errors.Is(err, endpoint.ErrNotAlive) {
```

Given this, I'll remove this lock as well!

jrajahalme committed:
policy: Cache prefixes separately

Change selector cache to return a netip.Prefix instead of
*net.IPNet. This avoids conversions down the line.

Store identity prefix for each CIDR identity separately in a sync.Map, so
that selector cache mutex is not needed to access them.

Rename GetNetsLocked as GetPrefix, as it is now returning a single
netip.Prefix and it does not need locking.

Suggested-by: Casey Callendrello <cdc@isovalent.com>
Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Avoid scanning all entries with auth rules

Use the port/proto trie index to exclude uninteresing keys from the scan
needed for propagating auth properties from more general rules to more
specific ones.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Propagate proxy port and auth type in added Auth L3/L4 entries

The added L3/L4 entries were created with 0 proxy port, while the proxy
port should be copied from the L4 rule.

Auth type should propagate from a L3 rule with protocol and wildcard port
when another L3-wildcard rule without auth type has the same protocol and
a specific port. In this case a new entry with the L3 and proto from the
L3-rule and the port (and the same protocol) from the L4-only rule should
be created.

Add unit testing to validate the above.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Copy L4-key instead of L3-only key

Copy the L4-key when creating new L3/L4 entries due to policy precedence
rules and copy the Identity from the L3-only key. This way the code is
simpler and less error prone.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Simplify by using mapStateMap directly

Use the type mapStateMap directly rather than via mapState when adding
l3l4 entries for Auth policies.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Remove ForEachAllow() and ForEachDeny() from MapState interface

Currently there are no users so it is a good time to remove these.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add visibility keys without ranging over all entries

Use the cidr index to find interesting items for visibility key injection
without ranging over all entries in the MapState.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

nathanjsweet commented on pkg/container/bitlpm/trie.go:
```diff
@@ -185,6 +185,10 @@ func (t *trie[K, T]) Descendants(prefixLen uint, k Key[K], fn func(prefix uint,
 			return
 		}
+		// currentNode is a leaf and has no children. Calling k.BitValueAt may overrun the key storage.
+		if currentNode.prefixLen >= t.maxPrefix {
```
When did you run into this? In theory, this should be impossible given that `Upsert` sets the prefixLen of any key to at most the maxPrefix of the Trie.

jrajahalme replied:
Ran into this with CIDR trie (`bitlpm.CIDRTrie`) use. Got array out of bounds for an IPv4 CIDR when the bit 32 was asked for here.

nathanjsweet commented on pkg/policy/l4.go:
```diff
@@ -813,6 +813,7 @@ func createL4Filter(policyCtx PolicyContext, peerEndpoints api.EndpointSelectorS
 
 	// already validated via L4Proto.Validate(), never "ANY"
+	// NOTE: "ANY" for wildcarded port/proto!
 	u8p, _ := u8proto.ParseProtocol(string(protocol))
```
Not quite yet, as L4Filter still converts ANY to the supported protocols. Unless a later commit in this PR changes this we should change that for fully wildcarded port/proto as it will save on space in the map.

jrajahalme replied:
This would be safe to do, if the datapath will limit the proto to known supported one. Otherwise, in theory, there could be a `proto` which does not have ports in the traditional sense and the datapath could misbehave, if there is a ANY proto with port 80, etc.

nathanjsweet replied:
Let's address in a different PR.

nathanjsweet reviewed (changes requested):
This PR is on the right track, but how CIDRs and identity are indexed to one another and compared needs overhaul. This PR optimizes CIDR comparisons, but at the expense of the gains that can be made by optimizing port protocol comparison, which is, by far, the more common case. These two things do not have to be mutually exclusive, but the insertion logic in this PR neglects the gains that could be made with more robust port-protocol comparison. It is also unclear to me why this PR  traverses all CIDR ancestors of a given policy identity, when most of them can be completely eliminated from comparison by a simple port-protocol traversal of the indices. A better approach would be to index the CIDRs by key and do simple lookups within a port-protocol traversal.

jrajahalme committed:
policy: Make mapstate modifiers private

Make mapstate modifiers private. This helps later when modifiers need to
have selector cache locked.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add cidr index to mapstate

Add CIDR index to mapStateMap alongside the ID map. This allows for more
efficient lookup of super/sub-set entries for deny rule insertion.

This requires passing Identities interface along so that the CIDR can be
gotten from the selector cache.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

nathanjsweet commented on pkg/policy/mapstate.go:
```diff
@@ -502,6 +502,38 @@ func (msm *mapStateMap) ForEachBroaderCIDRDescendant(key Key, prefixes []netip.P
+func (msm *mapStateMap) ForEachPortProtoAncestor(key Key, f func(Key, MapStateEntry) bool) {
+	msm.trie.Ancestors(key.PrefixLength(), key, func(prefix uint, lpmKey bitlpm.Key[Key], idSet IDSet) bool {
+		k := lpmKey.Value()
+		if k.TrafficDirection == key.TrafficDirection {
```
Traffic direction is already indexed by trie. There is no need to do this comparison.

jrajahalme replied:
Fixed.

nathanjsweet commented on pkg/policy/mapstate.go:
```diff
@@ -502,6 +502,38 @@ func (msm *mapStateMap) ForEachBroaderCIDRDescendant(key Key, prefixes []netip.P
+func (msm *mapStateMap) ForEachPortProtoDescendant(key Key, f func(Key, MapStateEntry) bool) {
+	msm.trie.Descendants(key.PrefixLength(), key, func(prefix uint, lpmKey bitlpm.Key[Key], idSet IDSet) bool {
+		k := lpmKey.Value()
+		if k.TrafficDirection == key.TrafficDirection {
```
See above comment.

jrajahalme replied:
Fixed, thanks!

nathanjsweet commented on pkg/policy/mapstate.go:
```diff
@@ -273,6 +284,224 @@ func (msm *mapStateMap) ForEach(f func(Key, MapStateEntry) bool) bool {
+		// k is the key from trie with 0'ed ID
+		k := lpmKey.Value()
+
+		// Descendants iterates over the key itself if it is in the trie, skip it
```
It can iterate over a key of another equal port-protocol. In which case they should both be inserted, but no shadow-key is necessary.

jrajahalme replied:
Comment was a bit wrong, it now reads:
```
// Descendants iterates over equal port/proto, caller expects to see only narrower keys so skip it
```

nathanjsweet commented on pkg/policy/mapstate.go:
```diff
@@ -273,6 +284,224 @@ func (msm *mapStateMap) ForEach(f func(Key, MapStateEntry) bool) bool {
+		k := lpmKey.Value()
+
+		// Descendants iterates over the key itself if it is in the trie, skip it
+		if k.PortProtoIsEqual(key) {
```
This check should live in the argument func. I also don't think this comment is right.

jrajahalme replied:
Yes, the comment was wrong. I think the code is better here as it is needed to implement the "Narrower" semantic and we also must exclude identical identities, which makes the function structurally different.

jrajahalme replied:
I'll have think again if I could name these functions better, though.

jrajahalme replied:
Moving this check on the port/proto to the argument function would cause unnecessary traversals of the IDs under this port/proto in the IDSet or the CIDR trie, which are needed to call the argument function. IMO we should not do these unnecessary traversals.

nathanjsweet replied:
We don't need to do it in this pr, but we should rename these methods.

nathanjsweet reviewed (approved):
Wow! Truly incredible work.

jrajahalme committed:
policy: Add deny rule test and benchmark

Add test and benchmark with a mix of deny and allow CIDR rules.

Add identity for each CIDR so that 'toMapState' has some work to do.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Use CIDR index instead of scanning all entries for deny

Use PortProto and CIDR tries instead of scanning all entries when
preserving precedence for deny rules.

On a test benchmark with 1000 random allow and deny CIDR rules each, this
speeds up MapState generation by >1800x, hopefully making deny policies
useful in practical situations.

Optional validator interface is added to mapState so that we can keep the
existing identityIsSupersetOf logic in unit tests as additional
validation that the ancestor and descendats logic is visiting only
expeted entries.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

tommyp1ckles commented on pkg/policy/selectorcache.go:
```diff
@@ -91,6 +79,8 @@ type userNotification struct {
 // SelectorCache caches identities, identity selectors, and the
 // subsets of identities each selector selects.
 type SelectorCache struct {
+	prefixMap lock.Map[identity.NumericIdentity, netip.Prefix]
```
Q: Is this use of sync.Map justified over just having a lock? See:  https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/sync/map.go;l=15-23

jrajahalme replied:
Yes, fits the case (1):

> // The Map type is optimized for two common use cases: (1) when the entry for a given
// key is only ever written once but read many times

tommyp1ckles commented on pkg/policy/resolve.go:
```diff
@@ -164,6 +162,34 @@ func (p *EndpointPolicy) Detach() {
 	p.selectorPolicy.removeUser(p)
 }
 
+func NewMapStateWithInsert() (MapState, func(k Key, e MapStateEntry)) {
```
nit: Could we add a function comment for this, with an explanation for the `insert` funct callback?

jrajahalme replied:
I plan to remove this function in a later change, but added this for now:

> // NewMapStateWithInsert returns a new MapState and an insert function that can be used to populate
// it. We keep general insert functions private so that the caller can only insert to this specific
// map.

tommyp1ckles replied:
thanks 🙏

tommyp1ckles commented on pkg/policy/mapstate.go:
```diff
@@ -1594,15 +1609,11 @@ func (ms *mapState) addVisibilityKeys(e PolicyOwner, redirectPort uint16, visMet
+	// Find entries with the same L4
+	ms.allows.ForEachKeyWithPortProto(key, func(k Key, v MapStateEntry) bool {
+		if k.Identity != 0 {
+			if v.ProxyPort == 0 {
```
nit: these can go on the same line

jrajahalme replied:
These are intentionally on different lines for that following code does not need to shift left. This is to make reviewing easier. We can later change this in a refactor-only PR.

tommyp1ckles reviewed (commented):
@cilium/endpoint changes look good, just a few small comments/questions.

tommyp1ckles reviewed (approved):
Approved for @cilium/endpoint
